### ISSUE TITLE
Do we really need the HTTP_* constants

### ISSUE THREAD

**Issue by nyamsprod** [lukasoppermann/http-status#12]
Going back to the point I made during the PR from @colinodell I'm still not convinced about the usefulness of all the `HTTP_*` constants. Again a class constant should have a meaning only in the class context which uses it.

The `HttpStatus::HTTP_*` constants are designed to be used **outside** of the `Httpstatus` class since the class in itself does not rely on their values or presences to do anything meaningful. I would say that in the context of your class the methods already exposed are just about enough. 

A simple example below to illustrate what I'm saying. 

``` php
$response = new HttpResponse();
$response->withStatusCode(Httpstatus::HTTP_NOT_FOUND);
```

If I were to use your class like this, It would mean that I have created a **hard dependency** to `Httpstatus` and that I am relying on a constant value on which I have no control over. If `Httpstatus::HTTP_NOT_FOUND` value were to change I would have to go back in all my codes to update them. 

Conversely I have no problem with the following code:

``` php
$response = new HttpResponse();
$response->withStatusCode(HttpResponse::HTTP_NOT_FOUND);
```

In this last example, the constant used belong to the class, the developer was smart enough to add them to ease creating his/her response message. The class knows about the constant and changing its value won't affect my code as long as the developer is consistent with himself in that class. I think this is what Symfony did.

What if the developer forgot about the 404 error response, The constants are then handy. Well no, because you do provide a solid documentation page that's point to the right RFC. This is a educational problem not a code one.

Bottom line, with the current public API I can already do a bunch of things without having to rely on the `HTTP_*` constants, removing them would make for a cleaner, more compact class without any subtantial tradeoff.

**Comment by lukasoppermann** [comment]
Hmm, I do see your point in that they should not be used outside. However I still like the idea of providing those constants as it makes for a much more readable code. Of course if everybody would just know all codes by hard we would not need it, but this is not realistic. `HTTP_TOO_MANY_REQUESTS` is much easier to understand than `429`.

Maybe a solution could be to provide an additional interface that has all constants, so one could implement this interface and thus have all constants available in something like a response class, etc.?

**Comment by nyamsprod** [comment]
Using a Interface would only move the problem to another layer It won't resolve it. I understand your point but the only way to make your constants worth their presences is if you add usefulness to them. Adding a method like `Httpstatus::getHeaderLine` **may** do the trick:

``` php
$httpstatus = new HttpStatus();
$httpStatus->getHeaderLine(HttpStatus::HTTP_TOO_MANY_REQUESTS, 1.1);
//returns something like "HTTP/1.1 429 Too Many Requests"
```

But then again this is a narrow view since adding new status codes does not trigger adding new  constants.

``` php
$httpstatus = new HttpStatus([418 => "I'm A Teapot"]);
$httpstatus->getHeaderLine(HttpStatus::HTTP_IM_A_TEAPOT, 1.1); // won't work 
$httpstatus->getHeaderLine(418, 1.1); // should/will always work 
$httpstatus->getHeaderLine(429 , 1.1); // should/will always work 
```

So again you're _feature_ looks more like a duplication at best. For interop, IMHO people will use the integer as they are well described and finding informations about http status code is quite easy on the web.

And as I understand you are not in favor of adding a `Httpstatus::getHeaderLine` method and I do understand the whys

**Comment by lukasoppermann** [comment]
No, adding something just to keep the constants in is not what I mean. :)

I think I misunderstand the "problem" you are referring to. I thought you were saying that constants should not be used outside the class, correct? So IMO an interface would solve it, as the user would implement the interface into his class and thus be able to do something like `MyResponse::HTTP_NOT_FOUND`.

Did I miss your point?

**Comment by nyamsprod** [comment]
Here's an alternative explanation of my issue :) http://stackoverflow.com/a/5354044/2316257

> If you write MyClass::FOO, you're hard-coded to the implementation details of MyClass. This creates a hard-coupling, which makes your code less flexible, and as such should be avoided. However, interfaces exist to permit exactly this type of coupling. Therefore MyInterface::FOO doesn't introduce any concrete coupling. With that said, I wouldn't introduce an interface just to add a constant to it.

So an Interface **may do** but using a interface just for declaring constants... its another story. I encourage you to read the full explanation. What's funny is that I haven't even research this before and I found the answer interesting :)

**Comment by lukasoppermann** [comment]
Hmm, well it is not that I desperately need them, I just don't see the harm. I would actually argue that  his statement is that constants are fine if the value does not change during the lifetime of the class. I doubt that the status codes will change. 

I did read him saying interfaces just for constants is evil, but he did not say why. What's your take on this?

And if you are afraid of locking yourself in you could just not use the interface. But if you favour verbose code over numbers they are there to be used.

**Comment by nyamsprod** [comment]
My main issue is that an interface defines an API not the value used by this API.

I'm not against a constant in a interface but I'm against giving a meaning to the value of that constant other than to differentiate it from another constant. For instance, you would never use the value of `PDO::FETCH_ASSOC` for anything. I'm sure you don't know the constant value but you do know when to use it to modify PDOStatement behavior.  

In your case `Httpstatus::HTTP_NOT_FOUND` is tied to the **404** so this mean that your interface constant is not just a constant anymore it contains it itself a hard coded value which is an implementation detail.

That's the problem, if changing the value of  `Httpstatus::HTTP_NOT_FOUND` can cause a problem to your implementation then you constant is not a constant, it is an implementation that has nothing to do in an interface.

**Comment by lukasoppermann** [comment]
Okay, sorry to pester you with this so much, but I still disagree at the moment. 

If you go for the [PHP.net](http://php.net/manual/en/language.constants.php) version of constants it reads like this:

> A constant is an identifier (name) for a simple value. As the name suggests, that value cannot change during the execution of the script (except for magic constants, which aren't actually constants).

It think `HTTP_NOT_FOUND` is a very good name for a simple (int) value `404` that will not change during the execution of the script (or at all as long as the RFC does not change). Of course, if you change the constant, it will break, but if you change `E_ALL` it would also break the logging. If you implement the interface and afterwards remove it your script won't work anymore, but it will also not work if you remove the class in itself. 

I think your assumption is wrong: 

> That's the problem, if changing the value of Httpstatus::HTTP_NOT_FOUND can cause a problem to your implementation then you constant is not a constant, it is an implementation that has nothing to do in an interface.

Because if this is true than the whole concept of constants is wrong. What is of course true is concerning interfaces, of course if you implement it and overwrite it, you might be right, but this would be the same for any kind of constant in an interface or for that case also if you implement a method wrongly. Also this class does not rely on the interface, so actually if you set HTTP_NOT_FOUND to 123 but expect it to be 404 it is just plain stupid. :)

**Comment by colinodell** [comment]
I generally agree with @nyamsprod's argument in most cases, including [the discussion we had regarding the `CLASS_` constants](https://github.com/lukasoppermann/http-status/pull/9#issuecomment-128381809).

However, in this particular case, the `HTTP_` constants are being used instead of "magic numbers" like `404` or `503`.  IMHO it's generally considered a best-practice to replace magic numbers with meaningful constants, so I therefore think the constants should be kept.

**Comment by lukasoppermann** [comment]
So @colinodell would you be in favour of moving those constants into an interface or leaving them in the class?

Its basically external constants `$httpstatus::HTTP_NOT_FOUND` vs. `MyClass::HTTP_NOT_FOUND` which the disadvantage of writing an interface for the sole purpose of storing constants.

While I accept that the interface one may not be the "suggested" way I did not really read any problem with it, apart from a "its not nice to do" argument.

**Comment by nyamsprod** [comment]
@lukasoppermann @colinodell I still feel _it's not nice to do this_. But If you still want to proceed with this, I would recommend putting them into a distinct interface. Why ? because according to [PHP.net](http://php.net/manual/en/language.oop5.interfaces.php#language.oop5.interfaces.constants): 

>  It's possible for interfaces to have constants. Interface constants works exactly like class constants except they cannot be overridden by a class/interface that inherits them.

**Comment by lukasoppermann** [comment]
@nyamsprod somehow I missed this, this makes an interface even better. The good thing about this is, that its an "opt-in". If you do not want it, just don't implement it. :)

**Comment by nyamsprod** [comment]
Before re-inveting the wheel just notice this library :) https://github.com/shrikeh/teapot/tree/master/src/Teapot/StatusCode/RFC woth checking

**Comment by lukasoppermann** [comment]
@nyamsprod actually this was mentioned to me on twitter today, but I feel this packages has a different aim, with the the HTTPExceptions. I feel that this httpstatus could be part of the package you highlighted though. :)

**Comment by colinodell** [comment]
> So @colinodell would you be in favour of moving those constants into an interface or leaving them in the class?

I don't have a strong preference either way.

> But If you still want to proceed with this, I would recommend putting them into a distinct interface. Why ? because according to PHP.net:
> 
> > Interface constants works exactly like class constants except they cannot be overridden by a class/interface that inherits them.

Good point.  Although... you could mark the `Httpstatus` class as `final` which would achieve the same result.

But like I said, I personally don't have a strong preference of where they go, as long as they go somewhere.

**Comment by lukasoppermann** [comment]
Okay, I will implement something using an interface as I feel it is more clean a solution.

**Comment by lukasoppermann** [comment]
Okay, I added an implementation https://github.com/lukasoppermann/http-status/pull/13 I would be very happy if you could review it. Thanks.
